### PR TITLE
[Catalog] Debug UI for showing touches and safeAreaInsets

### PR DESCRIPTION
**Pull request by samnm**
This PR replaces the triple-tap to show touches, with triple tap to display debug UI. Within the debug UI you can choose to show touches, show safeAreaInsets or not show the UI again until next restart.

### CODE REVIEW

File: catalog/MDCCatalog/MDCCatalogDebugAlert.swift [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+  fileprivate var numberOfRows: Int {
+    return settings.count + 1
+  }
+  fileprivate var dismissCellRow: Int {
+    return settings.count
+  }
+
+  override var preferredContentSize: CGSize {
+    get { return CGSize.init(width: 400, height: self.numberOfRows * 44) }
```
Comment (willlarche) [inline catalog/MDCCatalog/MDCCatalogDebugAlert.swift]: you don't need `.init`

File: catalog/MDCCatalog/MDCCatalogDebugAlert.swift [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    set { super.preferredContentSize = newValue }
+  }
+
+  required init?(coder aDecoder: NSCoder) {
```
Comment (willlarche) [inline catalog/MDCCatalog/MDCCatalogDebugAlert.swift]: You can mark this API as unavailable for this class.

File: catalog/MDCCatalog/MDCCatalogDebugAlert.swift [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    super.init(collectionViewLayout: layout)
+
+    self.transitioningDelegate = dialogTransitionController
+    self.modalPresentationStyle = .custom
+  }
+
+  override func viewDidLoad() {
+    super.viewDidLoad()
+
+    self.collectionView?.backgroundColor = UIColor.white
```
Comment (willlarche) [inline catalog/MDCCatalog/MDCCatalogDebugAlert.swift]: Just `.white`

File: catalog/MDCCatalog/MDCCatalogDebugAlert.swift [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "toggle", for: indexPath)
+    if let cell = cell as? MDCCatalogDebugToggleCell {
+      cell.setting = settings[indexPath.row]
+    }
+    return cell
+  }
+
+  override func collectionView(_ collectionView: UICollectionView,
+                               numberOfItemsInSection section: Int) -> Int {
+    return self.numberOfRows
```
Comment (willlarche) [inline catalog/MDCCatalog/MDCCatalogDebugAlert.swift]: Try removing all `self.`s and only add back in the ones Xcode forces you to.

File: catalog/MDCCatalog/MDCCatalogDebugAlert.swift [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+  override func layoutSubviews() {
+    super.layoutSubviews()
+
+    let width = self.contentView.bounds.width
+    let height = self.contentView.bounds.height
+
+    let switchSize = toggleSwitch.sizeThatFits(self.contentView.bounds.size)
+    toggleSwitch.frame = CGRect.init(x: width - switchSize.width - 10,
+                                     y: (height - switchSize.height)/2.0,
```
Comment (willlarche) [inline catalog/MDCCatalog/MDCCatalogDebugAlert.swift]: I think we do spaces before and after the `/`.

File: catalog/MDCCatalog/MDCCatalogDebugAlert.swift (hunk 2 above)
Comment (willlarche) [reply catalog/MDCCatalog/MDCCatalogDebugAlert.swift]: Goes for all of them.

File: catalog/MDCCatalog/MDCCatalogDebugAlert.swift (hunk 1 above)
Comment (willlarche) [reply catalog/MDCCatalog/MDCCatalogDebugAlert.swift]: Goes for all of them.

File: catalog/MDCCatalog/MDCCatalogWindow.swift [hunk 6]
```diff
@@ -108,5 +202,6 @@ class MDCTouchView: UIView {
     self.layer.cornerRadius = touchCircleSize / 2
     self.layer.borderColor = touchCircleBorderColor.cgColor
     self.layer.borderWidth = touchCircleBorderWidth
+    self.isUserInteractionEnabled = false
```
Comment (willlarche) [inline catalog/MDCCatalog/MDCCatalogWindow.swift]: You may not need `self`s in here.

**Review by willlarche — CHANGES_REQUESTED** [review]
Went thru for Swift style. It's good clean Swift!

File: catalog/MDCCatalog/MDCCatalogDebugAlert.swift (hunk 2 above)
Comment (samnm) [reply catalog/MDCCatalog/MDCCatalogDebugAlert.swift]: Prepended those methods with @available(*, unavailable).

File: catalog/MDCCatalog/MDCCatalogDebugAlert.swift (hunk 3 above)
Comment (samnm) [reply catalog/MDCCatalog/MDCCatalogDebugAlert.swift]: Done.

File: catalog/MDCCatalog/MDCCatalogDebugAlert.swift (hunk 4 above)
Comment (samnm) [reply catalog/MDCCatalog/MDCCatalogDebugAlert.swift]: DOne.

File: catalog/MDCCatalog/MDCCatalogWindow.swift (hunk 6 above)
Comment (samnm) [reply catalog/MDCCatalog/MDCCatalogWindow.swift]: Done.

File: catalog/MDCCatalog/MDCCatalogDebugAlert.swift (hunk 5 above)
Comment (samnm) [reply catalog/MDCCatalog/MDCCatalogDebugAlert.swift]: Done.

File: catalog/MDCCatalog/MDCCatalogWindow.swift [hunk 7]
```diff
@@ -58,28 +73,107 @@ class MDCCatalogWindow: MDCOverlayWindow {
...
+    let width = frame.width
+    let height = frame.height
+    let insetHeight = height - safeAreaInsets.top - safeAreaInsets.bottom
+
+    if edgeViews.count < 4 {
+      return
+    }
+
+    // top
+    edgeViews[0].frame = CGRect.init(x: 0, y: 0, width: width, height: safeAreaInsets.top)
```
Comment (jverkoey) [inline catalog/MDCCatalog/MDCCatalogWindow.swift]: Drop the `.init`'s here as well.

**Review by jverkoey — APPROVED** [review]
Hell yes! Screenshots pleaaase :)

**Comment by samnm** [comment]
It actually breaks on rotation so I'm just fixing that first :)

**Comment by jverkoey** [comment]
Also: let's definitely pull this out to its own library!

File: catalog/MDCCatalog/MDCCatalogWindow.swift [hunk 8]
```diff
@@ -58,28 +73,107 @@ class MDCCatalogWindow: MDCOverlayWindow {
...
+    }
+    if showTouches {
+      layoutEdgeInsetViews()
+    }
+  }
+
+  fileprivate func showEdgeInsetViews() {
+    removeEdgeInsetViews()
+
+    for _ in 0...4 {
```
Comment (ianegordon) [inline catalog/MDCCatalog/MDCCatalogWindow.swift]: This should be 0...3.

**Review by ianegordon — CHANGES_REQUESTED** [review]

File: catalog/MDCCatalog/MDCCatalogWindow.swift (hunk 8 above)
Comment (jverkoey) [reply catalog/MDCCatalog/MDCCatalogWindow.swift]: Truth - or `0..<4` or `1...4`.

Comment (samnm) [reply catalog/MDCCatalog/MDCCatalogWindow.swift]: Good to know!

**Comment by ianegordon** [comment]
+1 to MDFOverlayWIndow as its own component.

**Review by willlarche — APPROVED** [review]

File: catalog/MDCCatalog/MDCCatalogWindow.swift (hunk 8 above)
Comment (samnm) [reply catalog/MDCCatalog/MDCCatalogWindow.swift]: Fixed.

**Review by ianegordon — APPROVED** [review]

**Comment by ianegordon** [comment]
Please fix Xcode 8 build before merge.

**Pull request merged**